src/Microsoft.DotNet.Build.Tasks/NuGetPush.cs

```diff
@@ -0,0 +1,140 @@
+    /// https://github.com/NuGet/Home/issues/5224 requests retrying on hangs in NuGet.exe.
+    /// </summary>
+    public class NuGetPush : Task, ICancelableTask
+    {
+        [Required]
+        public string NuGetExePath { get; set; }
+
+        [Required]
+        public ITaskItem[] Inputs { get; set; }
+
+        [Required]
+        public string Source { get; set; }
+
+        public string ApiKey { get; set; }
+
+        public string Verbosity { get; set; } = "Detailed";
+
+        public int TimeoutPerPushInSeconds { get; set; } = TimeSpan.FromMinutes(5).Seconds;
+
+        public int RetriesPerPush { get; set; } = 5;
+
+        private bool _cancelRemainingPushes;
+
+        public void Cancel()
+        {
+            _cancelRemainingPushes = true;
+        }
+
+        public override bool Execute()
+        {
+            foreach (ITaskItem input in Inputs)
+            {
+                for (int i = 0; i < RetriesPerPush; i++)
+                {
+                    if (_cancelRemainingPushes)
+                    {
+                        break;
+                    }
+
+                    var args = new List<string>
+                    {
+                        "push", input.ItemSpec,
+                        "-Source", Source,
+                        "-Timeout", TimeoutPerPushInSeconds.ToString(),
+                        "-Verbosity", Verbosity
+                    };
+
+                    if (!string.IsNullOrEmpty(ApiKey))
+                    {
+                        args.Add("-ApiKey");
+                        args.Add(ApiKey);
+                    }
+
+                    string attemptMessage = $"(attempt {i + 1}/{RetriesPerPush})";
+
+                    Log.LogMessage(
+                        MessageImportance.Normal,
+                        $"Pushing '{input.ItemSpec}' {attemptMessage} ...");
+
+                    int exitCode = RunProcess(NuGetExePath, args);
+                    if (exitCode == 0)
+                    {
+                        Log.LogMessage(
+                            MessageImportance.Normal,
+                            $"Pushing '{input.ItemSpec}' {attemptMessage} Done.");
+
+                        break;
+                    }
+
+                    string message =
+                        $"Pushing '{input.ItemSpec}' {attemptMessage} FAILED: " +
+                        $"exit code {exitCode}";
+
+                    if (i + 1 == RetriesPerPush)
+                    {
+                        Log.LogError(message);
+                    }
+                    else
+                    {
+                        Log.LogMessage(message);
+                    }
+                }
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        private int RunProcess(string exe, IEnumerable<string> args)
+        {
+            using (var pushProcess = new Process
+            {
+                StartInfo = new ProcessStartInfo(NuGetExePath, string.Join(" ", args))
```

Comment:
Sorry missed this comment:

>An alternative to Process is to call into the NuGet methods directly from the task. I think it's better to keep the BuildTools NuGet package version separate, because we can be sure we're using what other MyGet customers are using and can be more flexible when asked to upgrade NuGet.exe as a potential way to fix a problem.

While that is reasonable we already do have a number of nuget dependencies in our tasks.